<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate local-ancestry-aware GWAS results — get.la.aware.score • LOCATOR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate local-ancestry-aware GWAS results — get.la.aware.score"><meta property="og:description" content="This function uses a glmmkin class object from the null GLMM to perform score tests for local-ancestry-aware association with genotypes in a GDS file .gds file."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LOCATOR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate local-ancestry-aware GWAS results</h1>

    <div class="hidden name"><code>get.la.aware.score.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function uses a glmmkin class object from the null GLMM to perform score tests for local-ancestry-aware association with genotypes in a GDS file .gds file.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get.la.aware.score</span><span class="op">(</span></span>
<span>  <span class="va">null.obj</span>,</span>
<span>  <span class="va">geno.file</span>,</span>
<span>  <span class="va">outfile</span>,</span>
<span>  <span class="va">LAC.file</span>,</span>
<span>  n_PC_used <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  id.type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"character"</span>, <span class="st">"integer"</span>, <span class="st">"numeric"</span><span class="op">)</span>,</span>
<span>  MAF.range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e-07</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>  miss.cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  missing.method <span class="op">=</span> <span class="st">"impute2mean"</span>,</span>
<span>  is.dosage <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  n.batch <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-null-obj">null.obj<a class="anchor" aria-label="anchor" href="#arg-null-obj"></a></dt>
<dd><p>a class glmmkin or class glmmkin.multi object, returned by fitting the null GLMMusing glmmkin.</p></dd>


<dt id="arg-geno-file">geno.file<a class="anchor" aria-label="anchor" href="#arg-geno-file"></a></dt>
<dd><p>the full name of a GDS file (including the suffix .gds), or an object of class SeqVarGDSClass.</p></dd>


<dt id="arg-outfile">outfile<a class="anchor" aria-label="anchor" href="#arg-outfile"></a></dt>
<dd><p>the output file name.</p></dd>


<dt id="arg-lac-file">LAC.file<a class="anchor" aria-label="anchor" href="#arg-lac-file"></a></dt>
<dd><p>the full name of a binary LAC file.</p></dd>


<dt id="arg-n-pc-used">n_PC_used<a class="anchor" aria-label="anchor" href="#arg-n-pc-used"></a></dt>
<dd><p>the dimension of PCs that will be included in analysis. Default is set to the number of dimension of PCs in the LAC.file.</p></dd>


<dt id="arg-id-type">id.type<a class="anchor" aria-label="anchor" href="#arg-id-type"></a></dt>
<dd><p>the class of sample IDs. Possible values are "character", "integer", and "numeric".</p></dd>


<dt id="arg-maf-range">MAF.range<a class="anchor" aria-label="anchor" href="#arg-maf-range"></a></dt>
<dd><p>a numeric vector of length 2 defining the minimum and maximum minor allele frequencies of variants that should be included in the analysis (default = c(1e-7, 0.5)).</p></dd>


<dt id="arg-miss-cutoff">miss.cutoff<a class="anchor" aria-label="anchor" href="#arg-miss-cutoff"></a></dt>
<dd><p>the maximum missing rate allowed for a variant to be included (default = 1, including all variants).</p></dd>


<dt id="arg-missing-method">missing.method<a class="anchor" aria-label="anchor" href="#arg-missing-method"></a></dt>
<dd><p>method of handling missing genotypes. Either "impute2mean" or "omit" (default = "impute2mean").</p></dd>


<dt id="arg-is-dosage">is.dosage<a class="anchor" aria-label="anchor" href="#arg-is-dosage"></a></dt>
<dd><p>a logical switch for whether imputed dosage should be used from a GDS infile (default = FALSE).</p></dd>


<dt id="arg-n-batch">n.batch<a class="anchor" aria-label="anchor" href="#arg-n-batch"></a></dt>
<dd><p>an integer for how many SNPs should be tested in a batch (default = 100). The computational time can increase dramatically if this value is either small or large. The optimal value for best performance depends on the user’s system.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>@format a dataframe with following components:</p><dl><dt>SNP</dt>
<dd><p>SNP name, as supplied in snps.</p></dd>

<dt>CHR</dt>
<dd><p>Chromosome, copied from .gds file.</p></dd>

<dt>POS</dt>
<dd><p>physical position in base pairs, copied from .gds file.</p></dd>

<dt>REF</dt>
<dd><p>reference allele, copied from .gds file.</p></dd>

<dt>ALT</dt>
<dd><p>alternate allele, copied from .gds file.</p></dd>

<dt>MISSRATE</dt>
<dd><p>number of individuals with non-missing genotypes for each SNP.</p></dd>

<dt>AF</dt>
<dd><p>ALT allele frequency for each SNP.</p></dd>

<dt>N</dt>
<dd><p>total sample size.</p></dd>

<dt>SCORE</dt>
<dd><p>the summary score of the effect allele.</p></dd>

<dt>VAR</dt>
<dd><p>the variance of the summary score.</p></dd>

<dt>PVAL</dt>
<dd><p>local-ancestry-aware GWAS p-values.</p></dd>


</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hanxiao Sun, Han Chen.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>






  </body></html>

